File: pygame/app/characters/mole.py
import pygame
import random
from characters.base.character import Character

class Mole(Character):
    def __init__(self, x, y, speed, gravity):
        # モグラの画像をロード
        images = [
            pygame.image.load('assets/characters/mole_running_start.png'),
            pygame.image.load('assets/characters/mole_running_end.png')
        ]
        super().__init__(images, x, y, speed, gravity)
        self.dig_timer = 0
        self.ai_timer = 0  # AIの行動タイマー
        self.direction = random.choice(["left", "right"])  # 初期方向をランダムに設定

    def update(self, map_instance, clock, TILE_SIZE):
        """モグラの行動ロジックを更新する"""
        self.ai_timer += clock.get_time()

        if self.ai_timer >= 1000:  # 1秒ごとに行動を切り替え
            self.ai_timer = 0
            self.direction = random.choice(["left", "right", "dig"])  # 左右移動または掘る

        if self.direction == "left" and not map_instance.check_collision(
            self.x, self.y, self.width, self.height, "left", self.speed
        ):
            self.x -= self.speed
            self.facing_left = True
        elif self.direction == "right" and not map_instance.check_collision(
            self.x, self.y, self.width, self.height, "right", self.speed
        ):
            self.x += self.speed
            self.facing_left = False
        elif self.direction == "dig":
            # 掘る動作を実行
            dig_x = self.x // TILE_SIZE
            dig_y = self.y // TILE_SIZE
            # map_instance.dig_tile(dig_x, dig_y, "down", self.width, self.height, self.speed)

        # 重力と地面判定
        self.on_ground, new_y = map_instance.is_on_ground(self.x, self.y, self.height)
        if self.on_ground:
            self.y = new_y
        else:
            self.y += self.gravity

        # アニメーション更新
        self.animation_timer += clock.get_time()
        if self.animation_timer >= 60:
            self.animation_timer = 0
            self.frame_index = (self.frame_index + 1) % len(self.images)

File: frontend_src/node/frontend/src/mole.ts
// Mole.ts
import { Character } from "./Character";
import { Map } from "./map";

export class Mole extends Character {
  private aiTimer: number;
  private direction: string;

  constructor(x: number, y: number, speed: number, gravity: number) {
    const images = [new Image(), new Image()];
    images[0].src = "assets/characters/mole_running_start.png";
    images[1].src = "assets/characters/mole_running_end.png";
    super(images, x, y, speed, gravity);

    this.aiTimer = 0;
    this.direction = Math.random() > 0.5 ? "left" : "right";
    this.width = images[0].width;
    this.height = images[0].height;
  }

  update(mapInstance: Map, deltaTime: number) {
    // AI タイマーの更新
    this.aiTimer += deltaTime;

    if (this.aiTimer >= 1000) {
      this.aiTimer = 0;
      const actions = ["left", "right"]; // "dig" を削除
      this.direction = actions[Math.floor(Math.random() * actions.length)];
    }

    // 動作の実行
    switch (this.direction) {
      case "left":
        if (!mapInstance.checkCollision(this.x, this.y, this.width, this.height, "left", this.speed)) {
          this.x -= this.speed;
          this.facingLeft = true;
        }
        break;
      case "right":
        if (!mapInstance.checkCollision(this.x, this.y, this.width, this.height, "right", this.speed)) {
          this.x += this.speed;
          this.facingLeft = false;
        }
        break;
    }

    // 重力の適用と地面の確認
    const [isOnGround, newY] = mapInstance.isOnGround(this.x, this.y, this.height);
    if (isOnGround) {
      this.y = newY; // 接地している場合、Y座標を修正
    } else {
      this.y += this.gravity; // 重力を適用
    }

    // アニメーションの更新
    this.animationTimer += deltaTime;
    if (this.animationTimer >= 60) {
      this.animationTimer = 0;
      this.frameIndex = (this.frameIndex + 1) % this.images.length;
    }
  }


  draw_on_game(ctx: CanvasRenderingContext2D, cameraX: number, cameraY: number) {
    const width = this.images[0].width
    const height = this.images[0].height
    super.draw(ctx, cameraX, cameraY, width, height);
  }

  draw_on_toppage(ctx: CanvasRenderingContext2D, x: number, y: number) {
    const width = 500
    const height = 500
    super.draw(ctx, x, y, width, height);
  }
}
